cloudnet/src/main/java/de/dytanic/cloudnet/template/install/ServiceVersionProvider.java

```diff
@@ -0,0 +1,322 @@
+
+public class ServiceVersionProvider {
+
+    private final Map<ServiceVersionType.InstallerType, ServiceVersionInstaller> installers = new HashMap<>();
+
+    private final Map<String, ServiceVersionType> serviceVersionTypes = new HashMap<>();
+
+    public ServiceVersionProvider() {
+        this.installers.put(ServiceVersionType.InstallerType.DOWNLOAD, new SimpleDownloadingServiceVersionInstaller());
+        this.installers.put(ServiceVersionType.InstallerType.BUILD, new ProcessingServiceVersionInstaller());
+    }
+
+    public boolean loadServiceVersionTypes(String url) throws IOException {
+        this.serviceVersionTypes.clear();
+
+        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
+        connection.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11");
+
+        connection.connect();
+
+        boolean success = false;
+
+        try (InputStream inputStream = connection.getInputStream()) {
+            JsonDocument document = new JsonDocument().read(inputStream);
+
+            if (document.contains("versions")) {
+                Collection<ServiceVersionType> versions = document.get("versions", TypeToken.getParameterized(Collection.class, ServiceVersionType.class).getType());
+
+                for (ServiceVersionType serviceVersionType : versions) {
+                    this.registerServiceVersionType(serviceVersionType);
+                }
+
+                success = true;
+            }
+        }
+
+        connection.disconnect();
+        if (success) {
+            return true;
+        }
+
+        for (ServiceVersionType serviceVersionType : this.createDefaultVersionTypes()) {
+            this.registerServiceVersionType(serviceVersionType);
+        }
+
+        return false;
+    }
+
+    public void registerServiceVersionType(ServiceVersionType serviceVersionType) {
+        this.serviceVersionTypes.put(serviceVersionType.getName().toLowerCase(), serviceVersionType);
+    }
+
+    public Optional<ServiceVersionType> getServiceVersionType(String name) {
+        return Optional.ofNullable(this.serviceVersionTypes.get(name.toLowerCase()));
+    }
+
+    public boolean canBuildVersion(ServiceVersion serviceVersion, ServiceVersionType.InstallerType installerType) {
+        return !installerType.requiresSpecificJavaVersionToExecute() || serviceVersion.canRunOn(JavaVersion.getRuntimeVersion());
+    }
+
+    public void installServiceVersion(ServiceVersionType serviceVersionType, ServiceVersion serviceVersion, ITemplateStorage storage, ServiceTemplate serviceTemplate) {
+        if (!this.canBuildVersion(serviceVersion, serviceVersionType.getInstallerType())) {
+            throw new IllegalStateException("Cannot run " + serviceVersionType.getName() + "-" + serviceVersion.getName() + "#" + serviceVersionType.getInstallerType() + " on " + JavaVersion.getRuntimeVersion().getName());
```

Comment:
We might want to add the min and max java versions (if available) to this message.